packages/peregrine/src/Toasts/useToastContext.js

```diff
@@ -0,0 +1,99 @@
+import React, { createContext, useContext, useReducer } from 'react';
+/**
+ * The state of the toast store.
+ * @typedef {Object} ToastState
+ * @property {Object} toasts - unordered map of id:toast
+ */
+const initialState = {
+    toasts: new Map()
+};
+
+const reducer = (prevState = initialState, action = {}) => {
+    const { type, payload } = action;
+
+    switch (type) {
+        case 'add': {
+            const nextToasts = new Map(prevState.toasts);
+            const prevToast = nextToasts.get(payload.id);
+
+            const isDuplicate = !!prevToast;
+            let timestamp = payload.timestamp;
+            if (isDuplicate) {
+                // If this is a _new_ duplicate toast we need to clear the
+                // previous timeout to prevent premature removal.
+                window.clearTimeout(prevToast.removalTimeoutId);
```

Comment:
@jimbo as we discussed I moved the references to previous state out of the action and into the reducer. But this is also undesirable as we are now have a side effect.